### PR TITLE
DEPR: DateOffset.apply

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

Gives us the flexibility to consolidate the logic in `__add__`

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -458,7 +458,17 @@ cdef class BaseOffset:
             FutureWarning,
             stacklevel=1,
         )
-        return self.apply(other)
+        return self._apply(other)
+
+    def apply(self, other):
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:463]: can you type & add a doc-string (followup ok)

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -458,7 +458,17 @@ cdef class BaseOffset:
...
+        # GH#44522
+        warnings.warn(
+            f"{type(self).__name__}.apply is deprecated and will be removed "
+            "in a future version. Use `offset + other` instead",
+            FutureWarning,
+            stacklevel=2,
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:469]: use the find_stack_level

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/offsets.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx:469]: IIRC find_stack_level doesn't play nicely with cython methods

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx:463]: will add to my next CLN branch

File: pandas/_libs/tslibs/offsets.pyx (hunk 2 above)
Comment (jreback) [reply pandas/_libs/tslibs/offsets.pyx:469]: oh interesting, ok

**Review by jreback — APPROVED** [review]

**Pull request merged**